### DISCUSSION TITLE
Why c# doesn't support params of different generic types?

### DISCUSSION THREAD

**Discussion by JustNrik** [dotnet/csharplang#4096] [Q&A] [answered]
We have 2 delegates, Action and Func that have a ton of overloads to support multiple parameters, but I'm no sure why don't we have an easier way to make delegates like this, for example, `delegate void Func<T, TResult>(T... params);` or something, where T are multiple types instead of a specific one and there can be as many params as we want. But I'm not sure if this was suggested before or if this is impossible to implement?

**Comment by HaloFour** [comment, marked as answer]
`Action` and `Func` aren't two delegates, there are 34 separate delegates, one for each arity of parameters.  Every delegate type can only have one signature.  With generics and generic type parameters it is possible to extend those signatures to support multiple simple use cases, but you can't use generics to change the nature of the parameter, such as making it `ref` or `params`.  That requires a different delegate type.

It does sound like you're suggesting something like variadic generic parameters, such as https://github.com/dotnet/roslyn/issues/5058.

**Reply by iam3yal** [reply]
> It does sound like you're suggesting something like variadic generic parameters, such as [dotnet/roslyn#5058](https://github.com/dotnet/roslyn/issues/5058).

One of the features I really wished C# had, could have been nice but given how we have so many types with the different parameters in the framework don't know how useful it is but probably to new codebases or sth.
